### ISSUE TITLE
Should not warn assigning anonymous type instance with trivial nullability differences

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#33577]
Should not warn on the following assignments:
```C#
#nullable enable

interface IOut<out T> { }

class Program
{
    static void F1<T>(T? x, T? y) where T : struct
    {
        if (x == null) return;
        var a = new { x, y };
        var b = new { x = y, y };
        a = b; // warning: mismatch
    }

    static void F2<T>(IOut<T> x, IOut<T?> y) where T : class
    {
        var a = new { F = x };
        var b = new { F = y };
        b = a; // warning: mismatch
    }
}
```

**Comment by jcouv** [comment]
Just to clarify, this is because anonymous types are readonly, so we can treat their "type parameters" as `out`.
